### ISSUE TITLE
[bitnami/redis-sentinel] The sentinel marks master as down, because of changing IP

### ISSUE THREAD

**Issue by xunto** [bitnami/containers#30116]
### Name and Version

bitnami/redis-sentinel:6.2.11

### What architecture are you using?

amd64

### What steps will reproduce the bug?

1. Environment:
`Docker Compose version v2.6.0`
```
Linux valentin-XPS-15-9510 5.19.0-35-generic #36~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Feb 17 15:17:25 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```
or
```
Darwin MacBook-Pro-Konstantin-2.local 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64 x86_64
```
2. With basic config:
```yaml
services:
  redis-sentinel:
    image: 'bitnami/redis-sentinel:latest'
    environment:
      - REDIS_MASTER_HOST=redis
      - REDIS_MASTER_PORT=6379
      - REDIS_MASTER_SET=service-manager
    ports:
      - '26379:26379'
    depends_on:
      - redis
    volumes:
      - sentinel_data:/bitnami

  redis:
    image: 'bitnami/redis:latest'
    restart: always
    environment:
      - ALLOW_EMPTY_PASSWORD=yes
      - REDIS_REPLICATION_MODE=master
    ports:
      - "6379:6379"
    volumes:
      - redis_data:/bitnami/redis/data

volumes:
  redis_data:
  sentinel_data:
```
3. After some time or after redis restart sentinel marks master as down:
```
+sdown master service-manager 172.19.0.2 6379
```

### What is the expected behavior?

For everything to work normally.

### What do you see instead?

master is marked as down and is removed from the list. Redis stops working after that. The application fails with error: "No slave (or master) found"

### Additional information

Commentary from another employee:

> The scripts initialize Sentinel by resolving the name stored in REDIS_MASTER_HOST and writing the IP to the Sentinel config. Subsequent resolutions are not performed on subsequent runs, but Docker assigns a different IP to the name stored in REDIS_MASTER_HOST each time a container is launched. This causes the Sentinel config to contain an outdated IP for the Redis host.

To sum up, the problem is related to docker changing Redis container IP, making sentinel lose connection. In addition, there seem to be more reasons for Redis to change IP (not just restarts). And because the config is stored in the volume, restarting the sentinel may not help either.

**Comment by carrodher** [comment]
It seems it is not an issue related to the Bitnami redis container image or Helm chart but about how the application or environment is being used/configured.

For information regarding the application itself, customization of the content within the application, or questions about the use of the technology or infrastructure; we highly recommend checking forums and user guides made available by the project behind the application or the technology.

That said, we will keep this ticket open until the stale bot closes it just in case someone from the community adds some valuable info.

**Comment by xunto** [comment]
@carrodher I think that the main problem is that redis-sentinel image documentation provides inherently not working examples, which are almost the same as the one I use. So I expect you to change the documentation to provide a working example.

P.S. Pease, point me out if my configuration is somehow different from the example given in README.

**Comment by carrodher** [comment]
Can you share what is the documentation you're following so we can reproduce the issue on our side?

**Comment by xunto** [comment]
@carrodher https://hub.docker.com/r/bitnami/redis-sentinel/ > Using a Docker Compose file

**Comment by haideralipunjabi** [comment]
I am facing this issue. Is there a fix available?

**Comment by juan131** [comment]
Hi everyone!

I agree with my colleagues @carrodher and @aoterolorenzo: this issue is a consequence of lacking production capabilities in the docker compose scenario. This problem is resolved in production environments (such as Kubernetes) by using services, which provide stable virtual endpoints with static IP(s) and make use of dynamic iptables rules to route traffic to the corresponding ephemeral pods (with ephemeral IPs).

The docker compose setups Bitnami maintains provide a quick way to launch & setup containerized applications & micro-services that require interacting with other components, so developers can easily start working with the applications and getting familiar with them. However, they're not intended for production.

**Comment by xunto** [comment]
> Hi everyone!
> 
> I agree with my colleagues @carrodher and @aoterolorenzo: this issue is a consequence of lacking production capabilities in the docker compose scenario. This problem is resolved in production environments (such as Kubernetes) by using services, which provide stable virtual endpoints with static IP(s) and make use of dynamic iptables rules to route traffic to the corresponding ephemeral pods (with ephemeral IPs).
> 
> The docker compose setups Bitnami maintains provide a quick way to launch & setup containerized applications & micro-services that require interacting with other components, so developers can easily start working with the applications and getting familiar with them. However, they're not intended for production.

True, yet we encounter the problem during development, making it harder to work on our features.

I think the best course of action here would be to provide instructions on how to configure a static IP network, during the development.

**Comment by shtoltz** [comment]
I have encountered this problem too. The worst part is that there is no obvious way to repeat the initial setup on container start without errors.

If you just delete `/bitnami/redis-sentinel/conf/sentinel.conf` file in an attempt to restart the initialization
https://github.com/bitnami/containers/blob/76f256525f25e35fc2d8f1f15f42d42946336aeb/bitnami/redis-sentinel/7.4/debian-12/rootfs/opt/bitnami/scripts/libredissentinel.sh#L211-L212

you will face the `The configuration file /opt/bitnami/redis-sentinel/etc/sentinel.conf is not writable` error.

That's because in the end of initialization the directory `/opt/bitnami/redis-sentinel/etc` is removed and replaced with a symlink to `/bitnami/redis-sentinel/conf`:
https://github.com/bitnami/containers/blob/76f256525f25e35fc2d8f1f15f42d42946336aeb/bitnami/redis-sentinel/7.4/debian-12/rootfs/opt/bitnami/scripts/libredissentinel.sh#L261-L262

So, to successfully re-run the initialization, you need to do the following steps:
1. Remove file `/bitnami/redis-sentinel/conf/sentinel.conf`
2. Remove symlink `/opt/bitnami/redis-sentinel/etc`
3. Copy default configs from `/opt/bitnami/redis-sentinel/etc.default` to `/opt/bitnami/redis-sentinel/etc`
4. Execute setup script


In a docker-compose file it will look something like this:
```yaml
services:

  redis:
    image: redis
    ports:
      - 6379:6379

  redis-sentinel:
    image: bitnami/redis-sentinel
    command: >
      bash -c "rm -f /bitnami/redis-sentinel/conf/sentinel.conf &&
               rm /opt/bitnami/redis-sentinel/etc
               cp -r /opt/bitnami/redis-sentinel/etc.default /opt/bitnami/redis-sentinel/etc
               /opt/bitnami/scripts/redis-sentinel/setup.sh &&
               /opt/bitnami/scripts/redis-sentinel/run.sh"
    environment:
      - REDIS_MASTER_HOST=redis
      - REDIS_MASTER_SET=service
    ports:
      - 26379:26379
    depends_on:
      - redis

```

It's dirty, but at least it works for now.
